## **DDR GROUP PROJECT:**

In [3]:
# Importing libraries BeautifulSoup from bs4 to parse file, requests for HTTP
from bs4 import BeautifulSoup, Tag
import requests

# Importing time to introduce. sleep() for delays to mimic human behaviour and re for regular expressions 
import time
import re
import os

# Connecting to mysql database
import mysql.connector

In [4]:
# Importing Selenium libraries for webdriver, Chrome, 'By' to find elements, 'Keys' to send special characters
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys

### ABOUT:

The healthwarehouse is an online pharmacy website, which lets you navigate over multiple categorised medicine such as Diabetic Suuplies, Over the counter and Home medicals. They have also divided based on prescription services. Additionally, within these sub categories they have classified the products to enhance the user interactivity. This list helps in checking the prices, discounts and even auto re-order.

### Opening the webPage and rejecting the cookies pop-up

The cookies pop-up always shows up whenever a user enters the website, which is then rejected by clicking on the 'Reject All' Button. This is done by specifying the element with .XPATH and text 'Reject All'. .click() is used to click on the element and start an interaction.

In [47]:
# Calling the chrome driver
driver = webdriver.Chrome()

# Setting delay of 4 seconds 
time.sleep(4)

# Navigating to google base url and setting delay for 4 seconds
driver.get('https://www.healthwarehouse.com/')
time.sleep(4)

# Selecting the element by specifying path and mentioning text in it to reject the cookies pop - up
healthware = driver.find_element(By.XPATH, '//button[text()="Reject All"]')
healthware.click()
time.sleep(5)

### Logging in 'healthwarehouse.com' as a registered user

Sign in to the website is done with the help of 'email' and 'password' created, after which the keys for email 'pemail' and password 'ppasswd' is sent. The continue button is then clicked to execute the actions.

In [48]:
# Finding the element for Sign In and then clicking to navigate to the Login Page
signIn = driver.find_element(By.XPATH, '//a[text()="Sign In"]')
signIn.click()
time.sleep(5)

# The block for entering email is found by passing the id in the code, after which the value is entered through send_keys()
emailInput = driver.find_element(By.ID, 'pemail')
emailInput.send_keys("ywxiang121@gmail.com")

time.sleep(5)
# Similarly it is done for password
passInput = driver.find_element(By.ID, 'ppasswd')
passInput.send_keys("bax422finalproject")

time.sleep(2)
# The Login button is then located and clicked on
login = driver.find_element(By.XPATH, '//*[@id="continue-button"]')
login.click()

# time.sleep() is introduced to put delays and intervals
time.sleep(5)


### **Navigating to over the counter tab, Clicking on the specific category, saving the iterated pages in .html file**


### **COUGH, COLD AND FLU:**

Clicking on the OTC tab with specific Cough & Cold medidines using .click(). The page displays a list of medicine product with their respective price, reviews, SKUs and various information. These all information could be stored as a product database and could provide useful insights about availabilty and price.

The page is navigated by selenium driver, after which content is stored in a html file. A directory is created via makedirs to store all the files in the sample. Adding condition for the same to check if it exists, otherwise create a new one. path.koin() is used yo specifiy the directory path with the file name.

In [12]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Cough, Cold & Flu'
healthware = driver.find_element(By.XPATH, '//a[text()="Cough, Cold & Flu"]')
healthware.click()
time.sleep(5)

# Creating a directory and checkinh if it exists
directory = 'ddrFinalProject'
if not os.path.exists(directory):
    os.makedirs(directory)
    
# Extracting the page content for first page
soup = BeautifulSoup(driver.page_source, 'html.parser')
#print(soup.prettify())
# Closing browser after 5 seconds with the help of .quit()
time.sleep(5)

# Saving filename as healthcare-page1.html
filepath = os.path.join(directory, 'healthcare-cold-page1.html')
with open(filepath, 'w', encoding='utf-8') as file:
    file.write(soup.prettify())
print('Saved')

Saved


### **INTERACTION WITH PAGE SORTING:**

**OBSERVATIONS:** 

The URL of the website changes when the page is changed back and forth.
1. We can trigger the page changes here by modifying only the url in browser's address bar. This could be done my modifying the variable **page**, while fetching the url and also limiting the products upto 20.
2. GET Request is made when the page changes according to the selection fro page 2-10. The GET Request then modifies the reponse accordingly which depends on all the parameters passed and reloads the page data.
3. The variable associated in the URL with sorting is **'sort'**. This changes to value- **bestseller**, **Newest**, **Alphabetical** if filter is changed.
4. Additionally, to change the tabs, path variable is changed such as /over-the-counter for OTC and /hair-care for haircare. The specifics within this gets add on such as /over-the-counter/cough-cold-flu.

In [13]:
headers = {'User-Agent': 'Mozilla/5.0'}

# Iterating through rest of 10 pages starting from 2nd page and saving in the html file
for page in range(2, 11):  # use 2-11
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-cold-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=2&sort=bestseller
Filename: healthcare-cold-page2.html
Saved page 2
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=3&sort=bestseller
Filename: healthcare-cold-page3.html
Saved page 3
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=4&sort=bestseller
Filename: healthcare-cold-page4.html
Saved page 4
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=5&sort=bestseller
Filename: healthcare-cold-page5.html
Saved page 5
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=6&sort=bestseller
Filename: healthcare-cold-page6.html
Saved page 6
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=7&sort=bestseller
Filename: healthcare-cold-page7.html
Saved page 7
url: https://www.healthwarehouse.com/over-the-counter/cough-cold-flu?limit=20&page=8&sort=best

### **PAIN AND FEVER:**

Clicking on the OTC tab with specific Pain & Fever. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable /pain-fever instead of /cough-cold-flu. If any specific is selected such as aspirin, it adds on to path variable with same parameters passed.

In [18]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Pain & Fever'
healthware = driver.find_element(By.XPATH, '//a[text()="Pain & Fever"]')
healthware.click()
time.sleep(5)

In [19]:
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 19 pages starting from 2nd page and saving in the html file
for page in range(1, 20):  
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-fever-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=1&sort=bestseller
Filename: healthcare-fever-page1.html
Saved page 1
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=2&sort=bestseller
Filename: healthcare-fever-page2.html
Saved page 2
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=3&sort=bestseller
Filename: healthcare-fever-page3.html
Saved page 3
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=4&sort=bestseller
Filename: healthcare-fever-page4.html
Saved page 4
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=5&sort=bestseller
Filename: healthcare-fever-page5.html
Saved page 5
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=6&sort=bestseller
Filename: healthcare-fever-page6.html
Saved page 6
url: https://www.healthwarehouse.com/over-the-counter/pain-fever?limit=20&page=7&sort=bestseller
Filename: healt

### **GAS RELIEF:**

Clicking on the OTC tab with specific Gas Relief. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable /gas-relief instead added to digestive-health. This specifies that gas-relief falls under the tab digestive-health.


In [25]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Gas Relief'
healthware = driver.find_element(By.XPATH, '//a[text()="Gas Relief"]')
healthware.click()
time.sleep(5)

In [26]:
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 3 pages starting from 2nd page and saving in the html file
for page in range(1, 4):  
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/digestive-health/gas-relief?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-digestive-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/digestive-health/gas-relief?limit=20&page=1&sort=bestseller
Filename: healthcare-digestive-page1.html
Saved page 1
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/gas-relief?limit=20&page=2&sort=bestseller
Filename: healthcare-digestive-page2.html
Saved page 2
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/gas-relief?limit=20&page=3&sort=bestseller
Filename: healthcare-digestive-page3.html
Saved page 3


### **ANTACID & HEARTBURN RELIEF:**

Clicking on the OTC tab with specific Antacid & Heartburn Relief. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable /over-the-counter/digestive-health/antacid-heartburn-relief.

In [27]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Antacid & Heartburn Relief'
healthware = driver.find_element(By.XPATH, '//a[text()="Antacid & Heartburn Relief"]')
healthware.click()
time.sleep(5)

In [28]:
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 5 pages starting from 2nd page and saving in the html file
for page in range(1, 6):  # use
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-antacid-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page=1&sort=bestseller
Filename: healthcare-antacid-page1.html
Saved page 1
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page=2&sort=bestseller
Filename: healthcare-antacid-page2.html
Saved page 2
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page=3&sort=bestseller
Filename: healthcare-antacid-page3.html
Saved page 3
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page=4&sort=bestseller
Filename: healthcare-antacid-page4.html
Saved page 4
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antacid-heartburn-relief?limit=20&page=5&sort=bestseller
Filename: healthcare-antacid-page5.html
Saved page 5


### **ANTIDIARRHEAL:**

Clicking on the OTC tab with specific Antidiarrheal. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable /over-the-counter/digestive-health/antidiarrheal.


In [49]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Antidiarrheal'
healthware = driver.find_element(By.XPATH, '//a[text()="Antidiarrheal"]')
healthware.click()
time.sleep(5)

In [50]:
directory = 'ddrFinalProject'
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 2 pages starting from 2nd page and saving in the html file
for page in range(1, 3):  
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/digestive-health/antidiarrheal?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-antidiarrheal-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antidiarrheal?limit=20&page=1&sort=bestseller
Filename: healthcare-antidiarrheal-page1.html
Saved page 1
url: https://www.healthwarehouse.com/over-the-counter/digestive-health/antidiarrheal?limit=20&page=2&sort=bestseller
Filename: healthcare-antidiarrheal-page2.html
Saved page 2


### **ALLERGY AND SINUS:**

Clicking on the OTC tab with specific Allergy & Sinus. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable /over-the-counter/allergy-sinus.

In [39]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Antidiarrheal'
healthware = driver.find_element(By.XPATH, '//a[text()="Allergy & Sinus"]')
healthware.click()
time.sleep(5)

In [40]:
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 10 pages starting from 2nd page and saving in the html file
for page in range(1, 11):
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-allergy-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("Filename:", filename)
    with open(filepath, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=1&sort=bestseller
Filename: healthcare-allergy-page1.html
Saved page 1
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=2&sort=bestseller
Filename: healthcare-allergy-page2.html
Saved page 2
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=3&sort=bestseller
Filename: healthcare-allergy-page3.html
Saved page 3
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=4&sort=bestseller
Filename: healthcare-allergy-page4.html
Saved page 4
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=5&sort=bestseller
Filename: healthcare-allergy-page5.html
Saved page 5
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=6&sort=bestseller
Filename: healthcare-allergy-page6.html
Saved page 6
url: https://www.healthwarehouse.com/over-the-counter/allergy-sinus?limit=20&page=

### **HAIR CARE:**

Clicking on the OTC tab with specific Hair Care. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable personal-care/hair-care. It can be noticed that the main tab has now changed to hair care, within the specifics of personal care. 


In [41]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Antidiarrheal'
healthware = driver.find_element(By.XPATH, '//a[text()="Hair Care"]')
healthware.click()
time.sleep(5)

In [42]:
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 9 pages starting from 2nd page and saving in the html file
for page in range(1, 9): 
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-haircare-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("filepath:", filename)
    with open(filename, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=1&sort=bestseller
filepath: healthcare-haircare-page1.html
Saved page 1
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=2&sort=bestseller
filepath: healthcare-haircare-page2.html
Saved page 2
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=3&sort=bestseller
filepath: healthcare-haircare-page3.html
Saved page 3
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=4&sort=bestseller
filepath: healthcare-haircare-page4.html
Saved page 4
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=5&sort=bestseller
filepath: healthcare-haircare-page5.html
Saved page 5
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=6&sort=bestseller
filepath: healthcare-haircare-page6.html
Saved page 6
url: https://www.healthwarehouse.com/personal-care/hair-care?limit=20&page=7&sort=bestseller
filepath: healthcare-hair

### **OINTMENTS:**

Clicking on the OTC tab with specific Ointments. The results are then extracted and saved in html files.

**OBSERVATIONS:** The url here changes with path variable personal-care/skin-care/creams-ointments.

In [51]:
# Clicking on 'Over the Counter'
healthware = driver.find_element(By.XPATH, '//a[text()="Over the Counter"]')
healthware.click()
time.sleep(5)

# Clicking on 'Antidiarrheal'
healthware = driver.find_element(By.XPATH, '//a[text()="Creams & Ointments"]')
healthware.click()
time.sleep(5)

In [52]:
directory = 'ddrFinalProject'
headers = {'User-Agent': 'Mozilla/5.0'}
# Iterating through rest of 14 pages starting from 2nd page and saving in the html file
for page in range(1, 15):  
    # Generalising the url, with limit per page as 20, page as a variable which can be changed and used to navigate to different pages
    # Page is sorted with bestseller
    url = f"https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page={page}&sort=bestseller"
    print("url:", url)
    # Adding sleep before sending request for each page
    time.sleep(7)
    driver.get(url)
    # Parsing the content and saving files based on page number
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    time.sleep(4)
    filename = f'healthcare-ointment-page{page}.html'
    filepath = os.path.join(directory, filename)
    print("filepath:", filename)
    with open(filename, 'w', encoding='utf-8') as file:
        file.write(soup.prettify())
    print(f'Saved page {page}')

url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=1&sort=bestseller
filepath: healthcare-ointment-page1.html
Saved page 1
url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=2&sort=bestseller
filepath: healthcare-ointment-page2.html
Saved page 2
url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=3&sort=bestseller
filepath: healthcare-ointment-page3.html
Saved page 3
url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=4&sort=bestseller
filepath: healthcare-ointment-page4.html
Saved page 4
url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=5&sort=bestseller
filepath: healthcare-ointment-page5.html
Saved page 5
url: https://www.healthwarehouse.com/personal-care/skin-care/creams-ointments?limit=20&page=6&sort=bestseller
filepath: healthcare-ointment-page6.html
Saved page 6
url: https://www

### **READING SAVED HTML FILES:**

1. Using the directory name, and iterating through each file in the specified directory.
2. Checking if the file ends with .html, then printing the list of files saved.

In [5]:
# Initialising directory name
directory = 'ddrFinalProject'

# Checking all html files with os.istdir()
# .endswith() check the extension
htmlFiles = [file for file in os.listdir(directory) if file.endswith(".html")]

# Printing the files
print("HTML files in the directory:")
for file in htmlFiles:
    print(file)

HTML files in the directory:
healthcare-haircare-page3.html
healthcare-fever-page14.html
healthcare-cold-page6.html
healthcare-fever-page2.html
healthcare-allergy-page2.html
healthcare-fever-page18.html
healthcare-ointment-page12.html
healthcare-ointment-page3.html
healthcare-ointment-page2.html
healthcare-ointment-page13.html
healthcare-fever-page19.html
healthcare-allergy-page3.html
healthcare-fever-page3.html
healthcare-cold-page7.html
healthcare-fever-page15.html
healthcare-antacid-page1.html
healthcare-haircare-page2.html
healthcare-ointment-page14.html
healthcare-ointment-page5.html
healthcare-fever-page8.html
healthcare-allergy-page8.html
healthcare-haircare-page5.html
healthcare-fever-page12.html
healthcare-antidiarrheal-page1.html
healthcare-digestive-page1.html
healthcare-fever-page4.html
healthcare-ointment-page9.html
healthcare-allergy-page4.html
healthcare-allergy-page5.html
healthcare-fever-page5.html
healthcare-ointment-page8.html
healthcare-cold-page1.html
healthcare-fe

### **EXTRACTING INFORMATION FROM HTML FILE:**

1. Parsing the specified directory 'ddrFinalProject', with iterating over all the files and opening/ reading if html file.
2. os.path.join() is used to specify the path of file in folder, while os.path.basname() specifies filename
3. os.path.isfile() is used to check if the filepath does specify a file or not
4. The file is then open with 'r' mode. 'r' is for reading mode.
5. Selecting the informations using select_one(), find_all(), select() methods and using .text to extract the texts field. .strip() is used to remove unwanted trailing or leading whitespaces.
6. The page is iterated with finding 'h3' which containts all the title in *<a>* anchor element. Iterating over each of the title, the information for single medicines is fetched.
7. If the item exists, title and url are selected. Other information are iterated by fetching the detailed page by clicking on the url. This page gives price, variant, description, image url link.
8. re.match(r"healthcare-(\w+)-page\d+\.html", filename), is used as a regex function to extract category from the file saved. While pushing files, we have saved it in this format. The middle word tells us about the tab. It specifies healthcare- followed by words followed by page(decimal number).html.
9. The data obtained is then stored in a dictionary which is then inserted into the list. This is done to store the data in a database. 

In [11]:
headers = {'User-Agent': 'Mozilla/5.0'}

# Mentioning the directory name
directory = 'ddrFinalProject'

# Creating a list to store all the data in the form of dictionaries
productDataList = []

# Iterating for all files in directory
for file in os.listdir(directory):
    # Joining the path with filename for accesing
    filepath = os.path.join(directory, file)
    # Printing the filename
    filename = os.path.basename(filepath)
    print(filename)
    # Checking if the file is html and if the filepath specifies a file type or not
    if os.path.isfile(filepath) and file.endswith(".html"):
        with open(filepath, 'r', encoding='utf-8') as file:
            # Reading the file and parsing the html content
            html = file.read()
            #print(html)
            soup = BeautifulSoup(html, 'html.parser')
            # Extracting all 'h3' items
            products = soup.find_all('h3') 
            # Initialising count
            count = 0
            # Iterating over each item
            for item in products:
                # Checking if item exists
                if item.a:
                    # Printing information
                    title = item.a.text.strip()
                    print('Count:', count)
                    print()
                    print('Title:', title)
                    count = count+1
                    url = item.a['href']
                    print("Link:", url)
                    # fetching url for detailed page, adding time.sleep() to mimic human behaviour
                    time.sleep(7)
                    page = requests.get(url, headers=headers)
                    soup = BeautifulSoup(page.content, 'html.parser')
                    #print(soup.prettify())
                    # Fetching the details, adding if condition to avoid exceptions
                    productId = soup.find('span', itemprop='sku')
                    if productId:
                        productId = productId.text.strip()
                        print("ProductId:", productId)
                    price = soup.find('span', itemprop='price')
                    if price:
                        price = price.text.strip()
                        print("Price:", price)
                    # Using regex function to extract the word in middle 
                    match = re.match(r"healthcare-(\w+)-page\d+\.html", filename)
                    if match:
                        # Fetching the first word in ()
                        category = match.group(1)
                        print("Category:", category)
                    variant = soup.find('h3', itemprop='brand')
                    if variant:
                        variantName = variant.find('a').text.strip()
                        print("Variant:", variantName)
                    availability = soup.find('h5') 
                    if availability:
                        availability = availability.text.strip()
                        print('Availability:', availability)
                    description = soup.find('h4', string='Description:')
                    if description:
                        # Extracting by specifying next sibling
                        para = description.find_next_sibling('p')
                        if para:
                            print("Description:", para.text.strip())
                    img = soup.find('img')
                    if img:
                        print("Image:", img['src'])

                    # Storing all the information in dictionary, later to be used for inserting in DB

                    productData = {
                        'title': title,
                        'url': url,
                        'productId': productId,
                        'price': price,
                        'category': category if category else None,
                        'variant': variantName if variantName else None,
                        'availability': availability if availability else None,
                        'description': description if description else None,
                        'image': img if img else None
                    }

                    # Removing keys with None values from the dictionary
                    productData = {key: value for key, value in productData.items() if value is not None}

                    # Adding the data dictionaries to the list
                    productDataList.append(productData)
 

healthcare-haircare-page3.html
Count: 0

Title: Mason Natural Coconut Oil Beauty Cream 2 oz.
Link: https://www.healthwarehouse.com/-2782755.html
ProductId: A10172585
Price: 8.90
Category: haircare
Availability: In Stock
Image: https://www.healthwarehouse.com/skin/frontend/hewa/newdesign/images/logo-top.png
Count: 1

Title: Vanicream Free & Clear Hair Styling Gel For Sensitive Skin 7oz
Link: https://www.healthwarehouse.com/-2740454.html
ProductId: A829205
Price: 13.81
Category: haircare
Availability: In Stock
Image: https://www.healthwarehouse.com/skin/frontend/hewa/newdesign/images/logo-top.png
Count: 2

Title: Just For Men Brush-In Mustache Beard & Sideburns Gel (Dark Brown) - 1oz
Link: https://www.healthwarehouse.com/-2732258.html
ProductId: A423614
Price: 15.75
Category: haircare
Availability: In Stock
Image: https://www.healthwarehouse.com/skin/frontend/hewa/newdesign/images/logo-top.png
Count: 3

Title: Herbal Essences Hello Hydration Conditioner - 11.7 fl oz
Link: https://www.hea

### **STORING THE DICTIONARY IN HTML FILE:**

1. This is done to ensure minimum data loss.
2. The starting of file is created by attaching the opening tags.
3. The products in the stored list is iterated and the information is then stored with format. Each product html is stored in main html.
4. The file is then closed with all the necessary tags, and is stored in the same directory.

In [13]:
# Making the file by attaching the starting of html file
htmlFile = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>
"""

# Iterating over the products in product list of dictionaries, creating div product to save it
for product in productDataList:
    productFile = f"""
    <div class="product">
        <h2>{product.get('title', '')}</h2>
        <img src="{product.get('image', '')}" alt="{product.get('title', '')}" style="width:100px;height:auto;">
        <p>Price: {product.get('price', '')}</p>
        <p>ID: {product.get('productId', '')}</p>
        <p>Category: {product.get('category', '')}</p>
        <p>Variant: {product.get('variant', '')}</p>
        <p>Availability: {product.get('availability', '')}</p>
        <p>Description: {product.get('description', '')}</p>
        <a href="{product.get('url', '#')}">Product Page</a>
    </div>
    <hr>
    """
    htmlFile += productFile
    # Appending product info to html file info

# Closing the tags
htmlFile += """
</body>
</html>
"""

# Writing it into the file 
# Mentioning the directory name
directory = 'ddrFinalProject'
file = "product_list.html"
filepath = os.path.join(directory, file)
with open(filepath, "w", encoding="utf-8") as file:
    file.write(htmlFile)

print(f"HTML file '{os.path.basename(filepath)}' has been created.")

HTML file 'product_list.html' has been created.


## **MONGO DB:**

### INITIALISING THE CONNECTION:

1. The Client is initialised using the port and localhost
2. db 'healthware' and collection 'products' is created in the specified client

In [26]:
from pymongo import MongoClient

# Creating a client with localhost and port: 27017
client = MongoClient('mongodb://localhost:27017')

# Creating database in client named 'healthware', collection within that database 'products'
db = client['healthware'] 
collection = db['products']

### INSERTING THE DATA

1. Converting the html file tag structure into urls or string value. This is done to ensure bson compatibility.
2. The dictionary is copied to ensure no change in original data
3. The elements value is checked for <tag>, if it has img name in it, the key is saved as a url. Otherwise, value is converted into string.

In [33]:
# Copying the dictionary
from copy import deepcopy
ProductDataListUpdated = deepcopy(productDataList)

# Iterating for reach product
for product in ProductDataListUpdated:
    # Iterating over each key value pair
    for key, value in product.items():
        # Checking if element is Tag, and then if it is img
        if isinstance(value, Tag):
            # src for images
            if value.name == 'img':
                product[key] = value['src']
            # If it is not image the value is converted into string, otherwise set as a url
            else:
                product[key] = str(value)

In [34]:
# Insert data to MongoDB

collection.insert_many(productDataList)
print("Data saved to MongoDB.")

Data saved to MongoDB.


## **BUSINESS INSIGHTS:**

In [1]:
from pymongo import MongoClient

client = MongoClient('mongodb://localhost:27017/?directConnection=true')
db = client['healthware'] 

### INSIGHT 1: AVERAGE PRICE OF ALL THE PRODUCTS
----------------------------------------------------

The average price for all results is calulated using the aggregate function. The result is grouped by to get a singular result. toDecimal is specified, before converting price to average for avoiding any exceptions. Next function fetches the first element. Since, the output consists of only one element, next() can be used.

**INFERENCE:** The average price turns out to be $14.18. However, this could vary depending on the quantity of medicines, also on they type such as strip, bottles etc.

In [2]:
# Fetching result after aggregating
result = db.products.aggregate([
    {"$group": {
        "_id": None,
        "averagePrice": {"$avg": {"$toDecimal": "$price"}}
    }}
])

avgeragePrice = next(result, {}).get('averagePrice', 'N/A')
avgeragePrice = str(avgeragePrice)
avgeragePriceRounded = round(float(avgeragePrice), 2)
print(f"Average Price: {avgeragePriceRounded}")

Average Price: 14.18


### INSIGHT 2: AVERAGE PRICE FOR ALL CATEGORIES
------------------------------------------------------

To deep dive into how it is seggregated within the categories, average price for each of them is checked. id is specified to which column we want to group. The todecimal converts it to decimal for computing average.

**INFERENCE:** Compared to the average price of all products, allergy medicines have a higher average price. Meanwhile, digestive medicines suggests a lower average price.  

In [3]:
# Find the average price for all categories
result = db.products.aggregate([
    {"$group": {
        "_id": "$category", # grouping by category
        "averagePrice": {"$avg": {"$toDecimal": "$price"}}
    }}
])

# Printing category and price
for data in result:
    avgeragePrice = str(data['averagePrice'])
    print(f"Category: {data['_id']}, Average Price: {round(float(avgeragePrice), 2)}")

Category: antacid, Average Price: 14.49
Category: antidiarrheal, Average Price: 12.17
Category: digestive, Average Price: 11.59
Category: fever, Average Price: 13.27
Category: haircare, Average Price: 13.5
Category: cold, Average Price: 14.87
Category: allergy, Average Price: 17.71
Category: ointment, Average Price: 13.53


### INSIGHT 3: CHEAPEST PRODUCT
------------------------------------

After finding the average, we wanted to see the cheapest product on OTC business. This is done by taking the price and sorting in ascending order. The first output is taken for the lowest price.

**INTERESTING FIND:** According to MedlinePlus.gov, nonprescription ibuprofen is utilized to lower fever and alleviate minor aches and pains stemming from various conditions including headaches, muscle aches, arthritis, menstrual periods, the common cold, toothaches, and backaches. The broad application and affordability of ibuprofen may lead to misuse and abuse. Besides, researchers have identified significant side effects associated with ibuprofen, such as heart attacks and gastrointestinal bleeding (Siemaszko, 2018). 

Linked URL - https://www.nbcnews.com/storyline/americas-heroin-epidemic/americans-are-abusing-over-counter-drugs-well-opioids-study-shows-n846401

In [6]:
# Finding the cheapest product
cheapestProduct = db.products.aggregate([
    {"$match": {"price": {"$exists": True}}}, # Checking if price exists
    {"$addFields": {"numeric_price": {"$toDecimal": "$price"}}}, 
    {"$sort": {"numeric_price": 1}}, # sorting in ascending
    {"$limit": 1} # getting the first value
])

for product in cheapestProduct:
    print(f"Product {product['title']} has the lowest price of {product['numeric_price']} in category - {product.get('category')}, it also has a variant of '{product.get('variant')}'")

Product Major Ibuprofen 200mg Tablets 24ct has the lowest price of 4.54 in category - fever, it also has a variant of 'Advil Tablets'


### INSIGHT 4: EXPENSIVE PRODUCT
-------------------------------------

Similarly, the last element is taken as the greatest value.

**INFERENCE:** The highest price is in category - allergy. This could be due to a very specific medicine for a rare allergy, or high dose.

In [7]:
# Finding the most expensive product
ExpensiveProduct = db.products.aggregate([
    {"$match": {"price": {"$exists": True}}},
    {"$addFields": {"numeric_price": {"$toDecimal": "$price"}}}, 
    {"$sort": {"numeric_price": -1}},
    {"$limit": 1}  
])

for product in ExpensiveProduct:
    print(f"Product {product['title']} has the highest price of {product['numeric_price']} in category - {product.get('category')}, it also has a variant of '{product.get('variant')}'")


ProductMucinex Expectorant Extended Release Tablet 600mg 500ct has the highest price of 258.14 in category -allergy, it also has a variant of 'Advil Capsules'


### INSIGHT 5: AVAILABILITIES AMONG CATEGORIES
-------------------------------------------

The analysis then focused on how the categories have availabilities across.

**RECOMMENDATION:** Cold category has the highest out-of-stock ratio. Given the prevalence and seasonal nature of colds, we recommend pharmacies optimize their inventory management.

In [13]:
# Calculating availabilities 
availabilityCategory = db.products.aggregate([
    {
        "$group": {
            "_id": {
                "category": "$category",
                "availability": "$availability"
            },
            "count": {"$sum": 1}  # Storing the count
        }
    }
])

# Initialising a dictionary
stockCategory = {}

# Iterating over each group
for group in availabilityCategory:
    # Extracting the information
    category = group["_id"]["category"]
    availability = group["_id"]["availability"]
    count = group["count"]

    # Initialising for the category key
    if category not in stockCategory:
        stockCategory[category] = {"In Stock": 0, "Out of Stock": 0}

    # Incrementing the count
    if availability == "In Stock":
        stockCategory[category]["In Stock"] += count
    elif availability == "Out of Stock":
        stockCategory[category]["Out of Stock"] += count

# Printing the results
for category, counts in stockCategory.items():
    print(f"CATEGORY: {category}, In Stock: {counts['In Stock']}, Out of Stock: {counts['Out of Stock']}")


CATEGORY: antacid, In Stock: 89, Out of Stock: 11
CATEGORY: haircare, In Stock: 145, Out of Stock: 15
CATEGORY: fever, In Stock: 347, Out of Stock: 31
CATEGORY: ointment, In Stock: 255, Out of Stock: 25
CATEGORY: digestive, In Stock: 60, Out of Stock: 0
CATEGORY: antidiarrheal, In Stock: 40, Out of Stock: 0
CATEGORY: cold, In Stock: 160, Out of Stock: 27
CATEGORY: allergy, In Stock: 199, Out of Stock: 1


### INSIGHT 7: VARIANT ANALYSIS
-----------------------------------

Finally, it was checked how variants for the products vary. Variants are the medicines that could be taken in place if the recommended medicine is out of stock. The analysis was done by grouping the products on variant and checking the number attached to it.

**INFERENCE:** Zyrtec, Voltaren, Extra Strength Tylenol are the top 3 variants that have the most product numbers.

In [14]:
# Grouping the products based on variants
variants = db.products.aggregate([
    {
        "$group": {
            "_id": "$variant",
            "count": {"$sum": 1}
        }
    },
    {
        "$sort": {"count": -1}  # Sorting on descending
    }
])

# Printing the results
for variant in variants:
    print(f"Variant: {variant['_id']}, Number of Products: {variant['count']}")


Variant: Zyrtec, Number of Products: 81
Variant: Voltaren, Number of Products: 63
Variant: Extra Strength Tylenol, Number of Products: 62
Variant: Tylenol 650mg Tablets, Number of Products: 58
Variant: Bacitracin Ointment, Number of Products: 55
Variant: Tylenol Children's Meltaways, Number of Products: 55
Variant: Advil Tablets, Number of Products: 52
Variant: Amlactin 12% Lotion, Number of Products: 50
Variant: Motrin 200mg Tablets White, Number of Products: 47
Variant: Imodium A-D Capsules, Number of Products: 42
Variant: Allegra, Number of Products: 41
Variant: Maalox Max Lemon Liquid, Number of Products: 40
Variant: Tylenol PM Capsules, Number of Products: 39
Variant: Benadryl® Allergy Ultratab, Number of Products: 39
Variant: Mucinex, Number of Products: 38
Variant: Prilosec, Number of Products: 37
Variant: Cortizone 5 Cream, Number of Products: 37
Variant: Cortizone 10+ Cream, Number of Products: 36
Variant: Lotrimin-AF Cream, Number of Products: 33
Variant: Neosporin Ointment, 

## **FINDINGS AND RECOMMENDATIONS:**

The data analysis suggested price variations across and within the categories. The price insights were done to find how this affects the people who have no insurance coverage. Insurance in the states plays a vital role in people having the access to care and even medicines. Due to surge in prices, public without insurance are not able to afford all the care that they want.

The analysis shows, that even though the average price is 14.18, the price can go upto 258.14 for an allergy medicine and even low to 4.54 for an advil tablet. Ibrufen being the lowest, also aids to the fact of *Americans abusing the over the counter medicines and especially advil substitutes*.

This could be due to various reasons such as this being a cheap and non prescriptive medicine treayed as a painkiller for multiple issues. Additionally, it also suggests that Advil has 52 number of products attached to it as variants. This also shows how vast the painkillers are kept. Even though tylenol has more variants, the price being low plays a huge role. We can also see that cold medicines have highest out of stock ratios, supporting the fact of usage of these medicines.

**RECOMMENDATIONS:** To sustain competitiveness in the market and enable strategy adaptation, pharma companies can have a constant monitoring plan for pricing changes and client preferences that might be put into place. It is also possible to comprehend the impact that insurance coverage has on individuals, emphasizing how prices may be changed and outliers can be identified. Communicating a wider range of medicines to customer, can help highlight the difference. Providing in stock/ out of stock availabilties, could help the stores in inventory management of most used medicines such as cold.

However, for consumers they could be educated on the misuse of ibrufen drugs for everything. Additionally promotion of variants with lower price could also be done to make the customer educated and not use the cheap OTC as a substitute. The main goal should be to have the same salt in medicine which could help in actually treating the illness rather than just subsiding the pain with ibrufen. T

